src/Cache/Context/OgMembershipStateCacheContext.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace Drupal\og\Cache\Context;
+
+use Drupal\Core\Cache\CacheableMetadata;
+use Drupal\Core\Cache\Context\CacheContextInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\og\GroupTypeManager;
+use Drupal\og\MembershipManagerInterface;
+use Drupal\og\OgMembershipInterface;
+
+/**
+ * Defines a cache context service, for "membership state" caching.
+ *
+ * Cache context ID: 'og_membership_state'
+ */
+class OgMembershipStateCacheContext implements CacheContextInterface {
+
+
+  /**
+   * The group type manager service.
+   *
+   * @var \Drupal\og\GroupTypeManager
+   */
+  protected $groupTypeManager;
+
+
+  /**
+   * The membership manager service.
+   *
+   * @var \Drupal\og\MembershipManagerInterface
+   */
+  protected $membershipManager;
+
+  /**
+   * The route match service.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $user;
+
+  /**
+   * Constructs a new UserCacheContextBase class.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The current user.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route match service.
+   * @param \Drupal\og\GroupTypeManager $group_type_manager
+   *   The group type manager service.
+   * @param \Drupal\og\MembershipManagerInterface $membership_manager
+   *   The membership manager service.
+   */
+  public function __construct(AccountInterface $user, RouteMatchInterface $route_match, GroupTypeManager $group_type_manager, MembershipManagerInterface $membership_manager) {
+    $this->user = $user;
+    $this->routeMatch = $route_match;
+    $this->groupTypeManager = $group_type_manager;
+    $this->membershipManager = $membership_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getLabel() {
+    return t('OG membership state');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getContext() {
+    if (!$route_contexts = $this->routeMatch->getRouteObject()->getOption('parameters')) {
+      // No "parameters" defined in the route.
+      return 'none';
+    }
```

Comment:
It is possible to add a 'runtimeContext' to the route (ref `ContextProviderInterface::getRuntimeContexts()`) - here we can dynamically set the right group that matches the current route. It should be possible to do this plugin based like was already being worked on.

We should use this runtimeContext to determine the cache context.